thirdparty/ipc/socket_windows.cpp

```diff
@@ -0,0 +1,187 @@
+{
+//    return 0;
+    unsigned long enable_non_blocking = 1;
+    return ioctlsocket(socket_handle, FIONBIO, &enable_non_blocking ) == 0;
+}
+int SocketImplementation::connect( int socket_handle, const struct sockaddr *address, socklen_t address_len )
+{
+    return ::connect(socket_handle, address, address_len);
+}
+int SocketImplementation::send( int socket_handle, const char * msg, size_t len)
+{
+    struct pollfd pfd;
+    pfd.fd = socket_handle;
+    pfd.events = POLLWRNORM;
+
+    if(SocketImplementation::poll(&pfd) == -1 )
+    {
+        SocketImplementation::perror("poll waiting error");
+        return -1;
+    }
+    else if( pfd.revents & POLLWRNORM )
+    {
+        printf("waiting for write [%d] %s \n", __LINE__, __FILE__);
+
+        int OK = ::send(socket_handle, msg, len, 0);
+        if(OK == -1)
+        {
+            SocketImplementation::perror("cant send message");
+            SocketImplementation::close(socket_handle);
+            return -1;
+        }
+        else
+        {
+            return 0;
+        }
+
+    }
+
+    return 0;
+}
+int SocketImplementation::recv( int socket_handle, char * buffer, size_t bufferSize )
+{
+    struct pollfd pfd;
+    pfd.fd = socket_handle;
+    pfd.events = POLLRDNORM;
+
+    if(SocketImplementation::poll(&pfd) == -1 )
+    {
+        SocketImplementation::perror("poll read error");
+        return -1;
+    }
+    else if( pfd.revents & POLLRDNORM )
+    {
+        printf("waiting for recv [%d] %s \n", __LINE__, __FILE__);
+
+        int OK = ::recv(socket_handle, buffer, bufferSize, 0);
+        if(OK == -1)
+        {
+            SocketImplementation::perror("cant read message");
+            SocketImplementation::close(socket_handle);
+            return -1;
+        }
+        else
+        {
+            return 0;
+        }
+    }
+
+    return 0;
+}
+
+int SocketImplementation::poll( struct pollfd * pfd, int timeout, int n )
+{
+    return ::WSAPoll(pfd, n, timeout);
+}
+
+int SocketImplementation::poll( int socket_handle )
+{
+    struct pollfd pfd;
+    pfd.fd = socket_handle;
+    pfd.events = POLLIN | POLLOUT;
+    pfd.revents = 0;
+    return ::WSAPoll(&pfd, 1, 0);
+}
+int SocketImplementation::accept( int socket_handle, struct sockaddr *addr, socklen_t * addrlen)
+{
+    return ::accept(socket_handle, addr, addrlen);
+}
+int SocketImplementation::bind( int socket_handle, const struct sockaddr *addr, size_t len)
+{
+    return ::bind(socket_handle, addr, len);
+}
+int SocketImplementation::listen( int socket_handle, int connection_pool_size)
+{
+    return ::listen(socket_handle, connection_pool_size);
+}
+void SocketImplementation::close( int socket_handle )
+{
+    ::closesocket(socket_handle);
+}
+void SocketImplementation::unlink( const char * unlink_file )
+{
+    _unlink(unlink_file);
+}
+
+void SocketImplementation::perror( const char * msg)
+{
+    LPVOID lpMsgBuf;
+    int e;
+
+    lpMsgBuf = (LPVOID)"Unknown error";
```

Comment:
There are other instances that are not sentence case.

Reply:
third-party lib